src/init/ClusterManager.ts

```diff
@@ -0,0 +1,184 @@
+      this.logger.warn(`Worker ${worker.process.pid} died with code ${code} and signal ${signal}`);
+      this.logger.warn('Starting a new worker');
+      cluster.fork().on('message', (msg: string): void => {
+        this.logger.info(msg);
+      });
+    });
+  }
+
+  /**
+   * Check whether the CSS server was booted in single threaded mode.
+   * @returns True is single threaded.
+   */
+  public isSingleThreaded(): boolean {
+    return this.clusterMode === ClusterMode.singleThreaded;
+  }
+
+  /**
+   * Whether the calling process is the primary process.
+   * @returns True if primary
+   */
+  public isPrimary(): boolean {
+    return cluster.isMaster;
+  }
+
+  /**
+   * Whether the calling process is a worker process.
+   * @returns True if worker
+   */
+  public isWorker(): boolean {
+    return cluster.isWorker;
+  }
+
+  /**
+   * Will throw an error if there are components instantiated implementing the {@link SingleThreaded}
+   * interface while the CSS is being run in multithreaded mode.
+   * @param componentsManager - The componentsManager being used to set up the application
+   */
+  public errorOnSingleThreadedComponents(componentsManager: ComponentsManager<any>): void {
+    const componentType = this.toComponentsJsType(componentsManager, 'SingleThreaded');
+    const violatingClasses: string[] = [];
+    for (const component of Object.values(componentsManager.componentResources)) {
+      if (component.isA(componentType) && component.value !== componentType) {
+        violatingClasses.push(component.value.split('#')?.[1]);
+      }
+    }
+    if (violatingClasses.length > 0) {
+      throw new InternalServerError(
+        `[${violatingClasses.join(',')}] is not threadsafe and should not be run in multithreaded setups!`,
+      );
+    }
+  }
+
+  private getFileName(): string {
+    return __filename;
+  }
+
+  /**
+   * Convert an exported interface name to the properly expected Components.js type URI.
+   * @param componentsManager - The currently used ComponentsManager
+   * @param localInterfaceName - An interface begin exported from the same source file as {@link ClusterManager}
+   * @returns A Components.js type URI
+   */
+  private toComponentsJsType(componentsManager: ComponentsManager<any>, localInterfaceName: string): string {
+    const mainPath = componentsManager.moduleState.mainModulePath;
+    if (!this.getFileName().startsWith(mainPath)) {
+      throw new InternalServerError(
+        `Could not convert ${localInterfaceName} to ComponentsJsType. (mainModulePath is incorrect)`,
+      );
+    }
+
+    const packageJson = componentsManager.moduleState.packageJsons[mainPath];
+    const lsdModule = packageJson['lsd:module'];
+    const lsdContextsKey = Object.keys(packageJson['lsd:contexts'])?.[0];
+    if (!lsdContextsKey.startsWith(lsdModule)) {
+      throw new InternalServerError(
+        `Could not convert ${localInterfaceName} to ComponentsJsType. (lsd:module is incorrect)`,
+      );
+    }
```

Comment:
If `mainPath` always needs to be the root file path of the CSS project you can use `modulePath()` from PathUtil instead of hoping the `mainModulePath` parameter is correct. This is how we find the package.json in the `ModuleVersionVerifier` for example.